### PR TITLE
[MXNET-342] Fix the multi worker Dataloader

### PR DESCRIPTION
**Pull request by ThomasDelteil**
## Description ##

[MXNET-342](https://issues.apache.org/jira/browse/MXNET-342)

Fix https://github.com/apache/incubator-mxnet/issues/9974

DataLoader was not compatible with ImageRecordDataset as the file descriptors ~~was (probably) closed on forking (close_fds=True by default)~~. 

The code in DataLoader use the `multiprocessing` package, which does not close the file descriptors, as it is simply calling `os.fork()`. The problem is that the recordIter is calling `lseek` on the file descriptors of each fork, all pointing to the same open file description. Since all the forked processes share the same open file description, they are all trying to set the value of `lseek` at the same time, thus creating a crash, as they can't be reading the same file at different position using a single open file description.

see https://stackoverflow.com/questions/4277289/are-file-descriptors-shared-when-forking
and https://stackoverflow.com/questions/11733481/can-anyone-explain-a-simple-description-regarding-file-descriptor-after-fork for more information of the distinction between `file descriptor` and `file description`

Now it reloads the record in each worker so that each individual process gets its own open file description.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by ThomasDelteil** [comment]
@piiswrong if you want to review

File: python/mxnet/gluon/data/dataloader.py [hunk 1]
```diff
@@ -112,6 +113,9 @@ def default_mp_batchify_fn(data):
 
 def worker_loop(dataset, key_queue, data_queue, batchify_fn):
     """Worker loop for multiprocessing DataLoader."""
+    if isinstance(dataset, RecordFileDataset):
```
Comment (piiswrong) [inline python/mxnet/gluon/data/dataloader.py]: don't do this kind of special casing.

Comment (ThomasDelteil) [reply python/mxnet/gluon/data/dataloader.py]: how do you suggest to do it then?

File: python/mxnet/gluon/data/dataset.py [hunk 2]
```diff
@@ -173,8 +173,15 @@ class RecordFileDataset(Dataset):
...
-        self._record = recordio.MXIndexedRecordIO(idx_file, filename, 'r')
+        self._filename = filename
+        self.reload_recordfile()
+
+    def reload_recordfile(self):
+        """
+        Reload the record file.
+        """
+        idx_file = os.path.splitext(self._filename)[0] + '.idx'
+        self._record = recordio.MXIndexedRecordIO(idx_file, self._filename, 'r')
```
Comment (piiswrong) [inline python/mxnet/gluon/data/dataset.py]: we should fix this in backend. Why would forking cause a problem? File descriptors should be duplicated when forking

Comment (ThomasDelteil) [reply python/mxnet/gluon/data/dataset.py]: That would be the ideal solution indeed.

File: python/mxnet/gluon/data/dataloader.py (hunk 1 above)
Comment (ThomasDelteil) [reply python/mxnet/gluon/data/dataloader.py]: @piiswrong we could also have a `reload` method on `Dataset` that does nothing for all dataset except for the RecordFileDataset ones, that way there is no special casing.

File: python/mxnet/gluon/data/dataset.py (hunk 2 above)
Comment (piiswrong) [reply python/mxnet/gluon/data/dataset.py]: isn't close_fds false by default?

Comment (ThomasDelteil) [reply python/mxnet/gluon/data/dataset.py]: For the subprocess package, it is `True` by default.

Comment (ThomasDelteil) [reply python/mxnet/gluon/data/dataset.py]: Ok digging a bit more, it seems that the `multiprocessing` package does not close file descriptors since it is simply calling `os.fork()`. I have updated the description of the PR to reflect the issue. tldr: file descriptors keep track of the position they are in the file. When forking, all process refer to the same file descriptor and try to move the current position at the same time, causing a crash.

**Comment by piiswrong** [comment]
this is a generic issue not specific to recordiodataset. Any dataset that opens a file could be affected. Does it behave the same way if the Dataset is written in python and opens file with `open`?

**Comment by ThomasDelteil** [comment]
It would behave the same way if the dataset relies on reading the file at run-time.

To make it clearer, instead of checking the `RecordIODataset`, we could have `RecordIODataset` inheriting from a new abstract class `FileReadingDataset` for example, that documents the behavior and has an abstract method `reload_file` that we call on the worker loop similarly as proposed in this PR.

**Comment by ThomasDelteil** [comment]
Just had a read through their code, I didn't see anything that would mitigate that issue. However they also don't provide a dataset that would read through a single file and use `lseek` to access records a la `.rec` file, so it is likely that they have never encountered that problem

**Comment by ThomasDelteil** [comment]
closing for now after @piiswrong design concerns. The bug is still present though.

**Comment by zhreshold** [comment]
I think pytorch is also suffering from similar problems: https://github.com/pytorch/pytorch/issues/973

I think we can use ForkingPickler.register(recordio.MXRecordIO, reopen_recordio) to force reload record files when workers are forked.
